feat(docs): Manual upgrade of the DB schema verison (#24)
vre-hub/vre-hub.github.io: docs/tech-docs/services/data-management.md

            nordugrid-arc-plugins-globus \
            nordugrid-arc-plugins-s3 \
            nordugrid-arc-plugins-xrootd
```

### Interacting with the Rucio Data Base 

> In general, Rucio database (DB) upgrades are straigforward and should work out of the box when following [Rucio official DB  documentation](https://rucio.github.io/documentation/operator/database/) or Rucio Major release notes. For instance, minor DB changes or updates will be applied when modifying the version of a chart.

Reviewer: ```suggestion
Generally, Rucio database (DB) upgrades are straigforward and should work out of the box when following [Rucio official DB documentation](https://rucio.github.io/documentation/operator/database/) or Rucio Major release notes. For instance, minor DB changes or updates will be applied when modifying the version of a chart.
```

#### Manual upgrade of the database schema

:::danger[User discretion advised]
Manual operations on the DB are dangerous. Please make sure you have a backup of your DB before continuing.
:::

> The following operations can used when getting some DB errors and not being able to upgrade the DB automatically (see some errors and examples [below](#rucio-conveyor-daemons-and-db-errors-examples)).

Reviewer: ```suggestion
The following actions can used when not being able to upgrade the DB automatically (see some errors and examples [below](#rucio-conveyor-daemons-and-db-errors-examples)).
```

1. Go to your dev environment and clone the rucio project.
2. Follow the [alembic tutorial](https://alembic.sqlalchemy.org/en/latest/tutorial.html#) and create a new `alembic.ini` file. 
```bash
git clone git@github.com:rucio/rucio.git
cd /rucio/lib/rucio/db/sqla
alembic init alembic
```
3. This command will create an `alembic.ini` file and an `alembic` subdirectory with the files and subdirectories described in the tutorial.
4. Edit the `alembic.ini` file adding the corresponding string to connect to the Rucio DB.
```bash
sqlalchemy.url = postgresql://rucio:PASSWORD@DBOD-CERN-URL.cern.ch:port/rucio
```
:::tip
Because we do not want to develop any new alembic version, modify the `alembic.ini` file and the `script_location` field to point the folder that contains the upstream rucio alembic revisions: `rucio/lib/rucio/db/sqla/migrate_repo/versions`
```bash
script_location = versions
```
:::

> 5. If you have enough experience with DBs, you can install the corresponding DB-client library to interact manually with it. Alternatively, you can also use UI applications to operate (e.g: DBeaver is free and open source).

Reviewer: ```suggestion
5. You can install the corresponding DB-client library to interact manually with it, or  alternatively (suggested for non experienced users), you can also use UI applications to operate (e.g: [DBeaver](https://dbeaver.io/) is free and open source).
```

```python
# Alembic revision identifiers
revision = '295289b5a800'
down_revision = 'a6eb23955c28'
```
When `grep`-ing for a specific reuvision, you will get two files as a result; the curret and the current plus DB revision.
```bash
> grep -ri 'a6eb23955c28'
295289b5a800_processed_by_and__at_in_requests.py:down_revision = 'a6eb23955c28'
a6eb23955c28_state_idx_non_functional.py:revision = 'a6eb23955c28'
```
Therefore, we can easily find the rev+1 file revision by doing 
```bash
> grep -ri 'a6eb23955c28' | grep "down_revision = 'a6eb23955c28'"
295289b5a800_processed_by_and__at_in_requests.py:down_revision = 'a6eb23955c28'
# and
> grep -ri 'a6eb23955c28' | grep "down_revision = 'a6eb23955c28'" | cut -d'_' -f1
295289b5a800
```

The following script can be used to find the order of the DB revisions starting from a `i-th` one.
```bash
#!/bin/bash
export I_REVISION=$1
export NEXT_REV=$I_REVISION
while [[ -n $NEXT_REV ]]; do 
    echo $NEXT_REV
    NEXT_REV=$(grep -ri ${NEXT_REV} | grep "down_revision = '${NEXT_REV}'" | cut -d'_' -f1)
done
```

> Once you discover the approximate problematic revision, you can play around with the script to understand the revision order. In our case we got

Reviewer: ```suggestion
Once you discover the problematic revision's identifier, you can play around with the script to understand the revision order. In our case we have
```

            nordugrid-arc-plugins-globus \
            nordugrid-arc-plugins-s3 \
            nordugrid-arc-plugins-xrootd
```

### Interacting with the Rucio Data Base 

In general, Rucio database (DB) upgrades are straigforward and should work out of the box when following [Rucio official DB  documentation](https://rucio.github.io/documentation/operator/database/) or Rucio Major release notes. For instance, minor DB changes or updates will be applied when modifying the version of a chart.

#### Manual upgrade of the database schema

:::warning
Entering danger zone. Please own under your own risk the following documentation.

> :::

Reviewer: ```suggestion
:::danger[User discretion advised]
Manual operations on the DB are dangerous. Please make sure you have a backup of your DB before continuing.
:::
```

│     cursor.execute(statement, parameters)
│ psycopg2.errors.UndefinedColumn: column requests.last_processed_by does not exist 
│ LINE 4: ...ND transfer_hops.next_hop_request_id IS NULL AND (requests.l...  
│                                                              ^       
│                                                                                                                      
```

and when trying to upgrade the DB, nothing was happening

```bash
[root@servers-rucio-server-8498f47bf9-cbq9j etc]# alembic upgrade head --sql
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Generating static SQL
INFO  [alembic.runtime.migration] Will assume transactional DDL.
BEGIN;

DROP TABLE alembic_version;

COMMIT;

[root@servers-rucio-server-8498f47bf9-cbq9j etc]# echo $(alembic current | cut -d' ' -f1)
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.

[root@servers-rucio-server-8498f47bf9-cbq9j etc]#
```

There was no trace of an `alembic.ini` file nowhere within our pods/VMs and no idea in which version we were.

##### How to solve this ?

> Trying not to diverge from Rucio practices as much as possible

Reviewer: ```suggestion

```

1. Go to your dev environment / VM and clone the rucio project.
2. Follow the [alembic tutorial](https://alembic.sqlalchemy.org/en/latest/tutorial.html#) and create a new `alembic.ini` file. 
```bash
git clone git@github.com:rucio/rucio.git
cd /rucio/lib/rucio/db/sqla
alembic init alembic
```
3. This command will create a `migrate-repo` directory as well. 
4. Add to the `alembic.ini` file the corresponding string to connect to the Rucio DB.
```
sqlalchemy.url = postgresql://rucio:PASSWORD@DBOD-CERN-URL.cern.ch:port/rucio
```

Also, installing the package to interact with a psql DB will be useful
```bash
dnf install postgresql-server
psql -h dbod-vre.cern.ch -U rucio -p 6600
\d distances
```

> At this point we spoke with an Rucio expert that pointed out that we were missing some table from a major DB migration around V32. We discovered that revision `295289b5a800` was not on our schema. But we did not know if previous revisions were applied too.

Reviewer: ```suggestion
```tip[Retrospective]
At this point we spoke with an Rucio expert that pointed out that we were missing some table from a major DB migration around V32. We discovered that revision `295289b5a800` was not on our schema. But we did not know if previous revisions were applied too.
```
```

Reviewer: Check the closing ticks please, not sure the suggestion mode worked this time.

When `grep`-ing on the directory with all the rucio revisions we  realise that we are getting two files, as any revision points to the curret one plus the previous one.
```bash
grep -ri 'a6eb23955c28'
295289b5a800_processed_by_and__at_in_requests.py:down_revision = 'a6eb23955c28'
a6eb23955c28_state_idx_non_functional.py:revision = 'a6eb23955c28'
```
Therefore, we can easily find the rev+1 file revision by doing 
```bash
grep -ri 'a6eb23955c28' | grep "down_revision = 'a6eb23955c28'"
295289b5a800_processed_by_and__at_in_requests.py:down_revision = 'a6eb23955c28'
# and
> grep -ri 'a6eb23955c28' | grep "down_revision = 'a6eb23955c28'" | cut -d'_' -f1
295289b5a800
```

The following snippet can be used to find the order of the DB revisions starting from a `i-th` one.
```bash
export I_REVISION=$1
export NEXT_REV=$I_REVISION
while [[ -n $NEXT_REV ]]; do 
    echo $NEXT_REV
    NEXT_REV=$(grep -ri ${NEXT_REV} | grep "down_revision = '${NEXT_REV}'" | cut -d'_' -f1)
done
```

> And by doing Obtaining `export I_REVISION=295289b5a800` and running the script we got

Reviewer: ```suggestion
By executing `export I_REVISION=295289b5a800`, and then running the script we got
```

```
a6eb23955c28
295289b5a800
27e3a68927fb
4df2c5ddabc0
a08fa8de1545
b5493606bbf5
b0070f3695c8
```

> By doing few `git checkout tags vX.Y.Z` and playing around wiht the output we got to

Reviewer: Clarify pls

```bash
fb28a95fe288 <-- revision 1.31.0
a6eb23955c28 <-- v32.0.0
295289b5a800
27e3a68927fb <-- last of v32.0.0
4df2c5ddabc0
a08fa8de1545 <-- Last of 34.6.0
b5493606bbf5
b0070f3695c8 <-- rucio main on Jan '25
```

Now we just need to check which revisions are applied to our DB by checking if the table(s) modification(s) are applied.

To allow `alembic` interact with all the revisions, it needed to be aware of them.
```
cd rucio/lib/rucio/db/sqla/migrate_repo
cp versions/* alembic/versions
```

and now we can force the alembic version on our DB.
```psql
psql -h dbod-vre.cern.ch -U rucio -p 6600

INSERT INTO alembic_version (version_num) VALUES ('140fef722e91') RETURNING alembic_version.version_num;
```
and now when doing
```psql
alembic current
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
140fef722e91
```
voilà !

> The rest was strightforwad, just applied the missing revisions until getting to the desired schema version.

Reviewer: ```suggestion
From now forward, it was sufficient to apply the missing revisions until getting to the desired schema version.
```

INFO  [alembic.runtime.migration] Running upgrade fb28a95fe288 -> a6eb23955c28, state idx non functional
> alembic upgrade 295289b5a800
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade a6eb23955c28 -> 295289b5a800, processed_by and _at in requests
> alembic upgrade 27e3a68927fb
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade 295289b5a800 -> 27e3a68927fb, remove replicas_tombstone and replicas_rse_id indexes
> alembic upgrade 4df2c5ddabc0
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade 27e3a68927fb -> 4df2c5ddabc0, remove temporary dids
> alembic upgrade a08fa8de1545
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade 4df2c5ddabc0 -> a08fa8de1545, transfer_stats table
```

> Getting successfully to v34.6.0 !

Reviewer: ```suggestion
Gotten successfully to v34.6.0 !
```
